src/tracer.c

```diff
@@ -1120,141 +1182,275 @@ static int __tracer_setup_cow_thread(struct snap_device *dev,
+ * original device's function pointer for handling i/o is used to process the BIO.
+ * Otherwise, depending on whether we're in snapshot or incremental mode, the appropriate 
+ * handler is called.
+ *
+ * Return: varies across versions of Linux and is what's expected by each for
+ *         a make request function.
+ */
+static asmlinkage MRF_RETURN_TYPE tracing_fn(struct bio *bio)
+
+#else
+
+/**     
+ * tracing_fn() - This is the entry point for in-flight i/o we intercepted.
  * @q: The &struct request_queue.
  * @bio: The &struct bio which describes the I/O.
  *
  * If the BIO has been marked as passthrough then the block device's
- * original MRF is used to process the BIO.  Otherwise, depending on
- * whether we're in snapshot or incremental mode, the appropriate handler
- * is called.
+ * original device's function pointer for handling i/o is used to process the BIO.
+ * Otherwise, depending on whether we're in snapshot or incremental mode, the appropriate 
+ * handler is called.
  *
  * Return: varies across versions of Linux and is what's expected by each for
  *         a make request function.
  */
-static MRF_RETURN_TYPE tracing_mrf(struct request_queue *q, struct bio *bio)
+static MRF_RETURN_TYPE tracing_fn(struct request_queue *q, struct bio *bio)
+
+#endif
 {
         int i, ret = 0;
-        struct snap_device *dev;
-        make_request_fn *orig_mrf = NULL;
-
+        struct snap_device *dev = NULL;
         MAYBE_UNUSED(ret);
 
         smp_rmb();
         tracer_for_each(dev, i)
         {
                 if (!dev || test_bit(UNVERIFIED, &dev->sd_state) ||
-                    !tracer_queue_matches_bio(dev, bio))
-                        continue;
+                        !tracer_queue_matches_bio(dev, bio))
+                                continue;
+                // If we get here, then we know this is a device we're managing
+                // and the current bio belongs to said device.
 
-                orig_mrf = dev->sd_orig_mrf;
-                if (dattobd_bio_op_flagged(bio, DATTOBD_PASSTHROUGH)) {
+                if (dattobd_bio_op_flagged(bio, DATTOBD_PASSTHROUGH))
+                {
                         dattobd_bio_op_clear_flag(bio, DATTOBD_PASSTHROUGH);
-                        goto call_orig;
                 }
-
-                if (tracer_should_trace_bio(dev, bio)) {
+                else if (tracer_should_trace_bio(dev, bio)) 
+                {
                         if (test_bit(SNAPSHOT, &dev->sd_state))
                                 ret = snap_trace_bio(dev, bio);
                         else
                                 ret = inc_trace_bio(dev, bio);
                         goto out;
                 }
-        }
-
-call_orig:
-        if (orig_mrf)
-                ret = dattobd_call_mrf(orig_mrf, q, bio);
-        else
-                LOG_ERROR(-EFAULT, "error finding original_mrf");
+                // Now we can submit the bio.
+                ret = SUBMIT_BIO_REAL(dev, bio);
```

Comment:
Yeah, that's definitively another problem w/ multiple partitions on the same disk. Looking into it...

Reply:
Ultimately, within that loop, the point of the tracer_queue_matches_bio check is to make sure the volume we're iterating is the same volume the bio belongs to, but this check is insufficient now that the queue is shared; I'm wondering if one potential solution could be to save and check the bi_partno in addition to the queue?

Reply:
How about just adding a check like ```if (dev->sd_base_dev->bd_partno != bio->bi_parno) continue;```

Reply:
Hi, @jnse 

I'm afraid that `bd_partno` is not supported in kernel v4.X or earlier.
Can we reuse partition’s sector range to check its partition number? ( `tracer_should_trace_bio` has already done this, maybe we can reuse it).


Besides, I guess we need to `goto out` after `SUBMIT_BIO_REAL` instead of continuing iterating to next `dev`. 

The problem is how to match the bio to corresponding snap_device exclusively

Reply:
Because my English is relatively poor.
I will try to write request later, please review it thanks!

Reply:
If my statement is not clear, I will 

1. try to write some change request later, please review.
2. make my statement clearer.

Thanks!

Best Regards,
Nick

Reply:
Hi Nick, your statements are clear,...
I think that bi_partno is still viable since we'd only be using it for the USE_BDOPS_SUBMIT_BIO case (so newer kernels).

Reply:
Hi @jnse 

We have 2 regressions needed to solve which are related to `bdev_get_queue`
1. `tracing_fn`
2. `__tracer_should_reset_mrf`
(I noticed that `find_orig_mrf` also uses `bdev_get_queue`, but it seems OK because it's only for before 5.8)

Reply:
Hi Nick! Thank you for all the reviews and finding issues. Please be patient with me ;) I've written up some code that fixes some of the above, but some issues came up while debugging with multiple partitions. While I can snapshot multiple partitions on the same drive now, writes to these partitions may 'hang', so I'm in the process of debugging what's up.